File: conformance/run_conformance.py
"""F3 cross-language conformance test — Python surface.

Loads ``conformance/vectors.json`` from the repo root and verifies that
``ogentic_redact._native.redact`` produces byte-identical output to the
expected values.  Any divergence is a pytest failure (→ CI red).

Run (from repo root, after ``maturin develop``):
    pytest conformance/run_conformance.py -v
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import pytest

# ── Import the Rust-backed binding ───────────────────────────────────────────

try:
    from ogentic_redact._native import (  # type: ignore[import]
        redact_with_salt as _redact_with_salt,
        unredact as _unredact,
    )
except ImportError as exc:
    pytest.skip(
        f"ogentic_redact._native not available — run `maturin develop` first. ({exc})",
        allow_module_level=True,
    )


# ── Load vectors ─────────────────────────────────────────────────────────────

_VECTORS_PATH = Path(__file__).parent.parent / "conformance" / "vectors.json"


def _load_doc() -> dict[str, Any]:
    with _VECTORS_PATH.open(encoding="utf-8") as fh:
        data = json.load(fh)
    assert data.get("vectors"), "vectors.json must contain at least one vector"
    assert data.get("call_salt_hex"), "vectors.json must carry a fixed call_salt_hex"
    return data


_DOC = _load_doc()
_VECTORS = _DOC["vectors"]
_SALT = bytes.fromhex(_DOC["call_salt_hex"])


# ── Parametrised test ─────────────────────────────────────────────────────────

@pytest.mark.parametrize("vector", _VECTORS, ids=[v["id"] for v in _VECTORS])
def test_f4_conformance_python(vector: dict[str, Any]) -> None:
    """Each vector must produce byte-identical text/tokens under the fixed salt,
    and round-trip back to the original input (ADR-0003 §9)."""
    result = _redact_with_salt(vector["input"], _SALT)

    assert result["text"] == vector["expected_text"], (
        f"[{vector['id']}] text mismatch\n"
        f"  input:    {vector['input']!r}\n"
        f"  got:      {result['text']!r}\n"
        f"  expected: {vector['expected_text']!r}"
    )
    assert result["tokens"] == vector["expected_tokens"], (
        f"[{vector['id']}] tokens mismatch\n"
        f"  input:    {vector['input']!r}\n"
        f"  got:      {result['tokens']!r}\n"
        f"  expected: {vector['expected_tokens']!r}"
    )

    restored = _unredact(result["text"], result["tokens"])
    assert restored == vector["input"], (
        f"[{vector['id']}] round-trip mismatch\n"
        f"  got:      {restored!r}\n"
        f"  expected: {vector['input']!r}"
    )

File: conformance/run_conformance.mjs
/**
 * F3 cross-language conformance test — Node.js surface.
 *
 * Loads `conformance/vectors.json` from the repo root and verifies that the
 * napi-rs `redact()` function produces byte-identical output to the expected
 * values.  Any divergence exits non-zero (→ CI red).
 *
 * Run (from repo root, after building the napi crate):
 *   node conformance/run_conformance.mjs
 *
 * The compiled `.node` addon is expected at:
 *   packages/ogentic-redact-node/ogentic_redact_node.linux-x64-gnu.node
 *   (or the platform-appropriate filename)
 */

import { createRequire } from 'module';
import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, '..');

// ── Load vectors ──────────────────────────────────────────────────────────────

const vectorsPath = resolve(repoRoot, 'conformance', 'vectors.json');
const { vectors } = JSON.parse(readFileSync(vectorsPath, 'utf-8'));

if (!vectors || vectors.length === 0) {
  console.error('vectors.json must contain at least one vector');
  process.exit(1);
}

// ── Load the napi-rs binding ──────────────────────────────────────────────────

const require = createRequire(import.meta.url);

// Try to find the compiled .node addon.  napi-rs names the file using the
// target triple, e.g. `ogentic_redact_node.linux-x64-gnu.node`.
let binding;
try {
  // napi-rs places the build artefact in the package directory after
  // `cargo build --release -p ogentic-redact-node` + napi-build postprocess.
  const pkgDir = resolve(repoRoot, 'packages', 'ogentic-redact-node');
  const { globSync } = await import('glob').catch(() => null) ?? {};
  let addonPath;
  if (globSync) {
    const matches = globSync(`${pkgDir}/*.node`);
    addonPath = matches[0];
  } else {
    // Fallback: conventional filename on linux-x64
    addonPath = resolve(pkgDir, 'ogentic_redact_node.linux-x64-gnu.node');
  }
  if (!addonPath) throw new Error('no .node file found');
  binding = require(addonPath);
} catch (err) {
  console.error(`[SKIP] Cannot load ogentic-redact-node binding: ${err.message}`);
  console.error('       Build it first: cargo build --release -p ogentic-redact-node');
  process.exit(0); // skip, not fail, when the addon is not yet built
}

const { redact } = binding;

// ── Run vectors ───────────────────────────────────────────────────────────────

let passed = 0;
let failed = 0;

for (const v of vectors) {
  const result = redact(v.input);

  const textOk = result.text === v.expected_text;
  const tokensOk = JSON.stringify(sortedObj(result.tokens)) === JSON.stringify(sortedObj(v.expected_tokens));

  if (textOk && tokensOk) {
    console.log(`  ✓  ${v.id}`);
    passed++;
  } else {
    console.error(`  ✗  ${v.id}`);
    if (!textOk) {
      console.error(`       text\n         got:      ${JSON.stringify(result.text)}\n         expected: ${JSON.stringify(v.expected_text)}`);
    }
    if (!tokensOk) {
      console.error(`       tokens\n         got:      ${JSON.stringify(result.tokens)}\n         expected: ${JSON.stringify(v.expected_tokens)}`);
    }
    failed++;
  }
}

console.log(`\n${passed} passed, ${failed} failed (Node.js surface, ${vectors.length} vectors)`);
process.exit(failed > 0 ? 1 : 0);

function sortedObj(obj) {
  return Object.fromEntries(Object.entries(obj).sort(([a], [b]) => a.localeCompare(b)));
}
